plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php

```diff
@@ -1823,23 +1831,40 @@ public function delete( &$order, $args = array() ) {
 	}
 
 	/**
-	 * Helper method to set child orders to the parent order's parent.
+	 * Set the parent id of child orders to the parent order's parent if the post type
+	 * for the order is hierarchical, just delete the child orders otherwise.
 	 *
 	 * @param \WC_Abstract_Order $order Order object.
 	 *
 	 * @return void
 	 */
-	private function upshift_child_orders( $order ) {
+	private function upshift_or_delete_child_orders( $order ) {
 		global $wpdb;
-		$order_table  = self::get_orders_table_name();
-		$order_parent = $order->get_parent_id();
-		$wpdb->update(
-			$order_table,
-			array( 'parent_order_id' => $order_parent ),
-			array( 'parent_order_id' => $order->get_id() ),
-			array( '%d' ),
-			array( '%d' )
-		);
+
+		$order_table     = self::get_orders_table_name();
+		$order_parent_id = $order->get_parent_id();
+
+		if ( $this->legacy_proxy->call_function( 'is_post_type_hierarchical', $order->get_type() ) ) {
+			$wpdb->update(
+				$order_table,
+				array( 'parent_order_id' => $order_parent_id ),
+				array( 'parent_order_id' => $order->get_id() ),
+				array( '%d' ),
+				array( '%d' )
+			);
+		} else {
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+			$child_order_ids = $wpdb->get_col(
```

Comment:
I think we'd need to call `$order->delete` here instead of immediately deleting, so that appropriate delete hooks gets fired.

Reply:
Ok, done. I had to implement the `get/set_verify_parent_id` methods in the order object too so that "Invalid parent ID" errors don't pop up.

Reply:
unit tests you added passes for me even I remove the call to `set_verify_parent_id`, can you also add a test that captures this behavior.